contracts/token/ERC7984/extensions/ERC7984Rwa.sol

```diff
@@ -0,0 +1,181 @@
+        _pause();
+    }
+
+    /// @dev Unpauses contract.
+    function unpause() public virtual onlyAdminOrAgent {
+        _unpause();
+    }
+
+    /// @dev Returns true if has admin role, false otherwise.
+    function isAdmin(address account) public view virtual returns (bool) {
+        return hasRole(DEFAULT_ADMIN_ROLE, account);
+    }
+
+    /// @dev Returns true if agent, false otherwise.
+    function isAgent(address account) public view virtual returns (bool) {
+        return hasRole(AGENT_ROLE, account);
+    }
+
+    /// @dev Adds agent.
+    function addAgent(address account) public virtual {
+        _addAgent(account);
+    }
+
+    /// @dev Removes agent.
+    function removeAgent(address account) public virtual {
+        _removeAgent(account);
+    }
+
+    /// @dev Mints confidential amount of tokens to account with proof.
+    function confidentialMint(
+        address to,
+        externalEuint64 encryptedAmount,
+        bytes calldata inputProof
+    ) public virtual returns (euint64) {
+        return _confidentialMint(to, FHE.fromExternal(encryptedAmount, inputProof));
+    }
+
+    /// @dev Mints confidential amount of tokens to account.
+    function confidentialMint(address to, euint64 encryptedAmount) public virtual returns (euint64) {
+        return _confidentialMint(to, encryptedAmount);
+    }
+
+    /// @dev Burns confidential amount of tokens from account with proof.
+    function confidentialBurn(
+        address account,
+        externalEuint64 encryptedAmount,
+        bytes calldata inputProof
+    ) public virtual returns (euint64) {
+        return _confidentialBurn(account, FHE.fromExternal(encryptedAmount, inputProof));
+    }
+
+    /// @dev Burns confidential amount of tokens from account.
+    function confidentialBurn(address account, euint64 encryptedAmount) public virtual returns (euint64) {
+        return _confidentialBurn(account, encryptedAmount);
+    }
+
+    /// @dev Forces transfer of confidential amount of tokens from account to account with proof by skipping compliance checks.
+    function forceConfidentialTransferFrom(
+        address from,
+        address to,
+        externalEuint64 encryptedAmount,
+        bytes calldata inputProof
+    ) public virtual returns (euint64) {
+        return _forceConfidentialTransferFrom(from, to, FHE.fromExternal(encryptedAmount, inputProof));
+    }
+
+    /// @dev Forces transfer of confidential amount of tokens from account to account by skipping compliance checks.
+    function forceConfidentialTransferFrom(
+        address from,
+        address to,
+        euint64 encryptedAmount
+    ) public virtual returns (euint64 transferred) {
+        return _forceConfidentialTransferFrom(from, to, encryptedAmount);
+    }
+
+    /// @dev Internal function which adds an agent.
+    function _addAgent(address account) internal virtual onlyAdminOrAgent {
+        _grantRole(AGENT_ROLE, account);
+    }
```

Comment:
I prefer to have the gating be done on the public function over the internal function. This is how we generally do it throughout the libs.